Fix complex matrix support in normestinv

marcusps opened the pull request:
Following discussion in Higham and Tisseur's paper.

_Nicholas J. Higham and Françoise Tisseur_ (2000) A block algorithm for matrix 1-norm estimation, with an application to 1-norm pseudospectra. **SIAM Journal On Matrix Analysis And Applications**, 21 (4). pp. 1185-1201. ISSN 1095-7162 ([preprint](http://eprints.ma.man.ac.uk/321/))

CC @mfasi @andreasnoack

marcusps commented:
Coincidentally, I was implementing Higham and Tisseur's more general norm estimation function, since it is needed in sparse matrix exponentiation (I started translating Al-Mohy and Higham's `expmv` routine), and came across this recent commit.

On a related note, `expmv` actually needs a slightly different version of the 1 norm estimation. It needs to estimate the norm of some power of the given matrix, and it needs to know how many iterations the estimation algorithm used (it uses this number to estimation needed accuracy).  

I don't imagine this more general function is really needed in base, but it does lead to some redundancy (the code is nearly identical, but instead of using `factorize(A)` and `\`, I'd use some in place sparse matrix multiplications.

Is it worth it to have the more general function in base, and have `normestinv` be a specialization of the more general function?

kshyatt commented:
Do H&T provide an example/check? Would it be possible to include a test for `complex` case?

tkelman commented:
How much more code are we talking about? If it's not exported, and/or well-tested and documented then it's probably okay to put in base if there's already closely-related functionality here.

trick for easier review: https://github.com/JuliaLang/julia/pull/12592/files?w=1

marcusps commented:
@kshyatt At the bottom of p. 1191 and and in the middle of p. 1195 the talk about complex matrices in particular.  The minor changes I pushed should cover everything (really, it is about the one line replacing `sign`, skipping the column randomization probably does not save all that much). I'll add some tests.

@tkelman It boils down to a couple of lines of code. I'll make a separate pull request for that.

kshyatt commented:
@marcusps Cheers, thanks! I love having more tests :).

mfasi commented:
@marcusps Thank you for fixing `normestinv`. Two thoughts:
* I think that skipping the randomization is a good improvement. On the other hand, I'd rather stick to `sign`, which I think computes the same quantity `x/abs(x)` for `x` complex and is also type safe (try for example `x = 1; x/abs(x), sign(x)`).
* I think that adding a more general `normest` function capable of estimating `norm(A^p,1)` without explicitly computing `A^p` would be very useful. If `A` is large enough, using it could improve the performance of `logm(A)` and `A^p` for instance.

StefanKarpinski commented on base/sparse/linalg.jl:
```diff
@@ -585,39 +585,41 @@ function normestinv{T}(A::SparseMatrixCSC{T}, t::Integer = 2)
         for j = 1:t
             for i = 1:n
-                S[i,j] = Y[i,j]==0?1:sign(Y[i,j])
+                S[i,j] = Y[i,j]==0?1:(Y[i,j]/abs(Y[i,j]))
```
`sign(z::Complex)` is already defined this way, but is more efficient for non-complex types.

marcusps replied:
Ah, I didn't realize sign was defined for complex numbers. The change then
boils down to the if statement :) I'll revert to the call to sign() and add
some tests.

On Wed, Aug 12, 2015, 6:17 PM Stefan Karpinski <notifications@github.com>
wrote:

> In base/sparse/linalg.jl
> <https://github.com/JuliaLang/julia/pull/12592#discussion_r36921191>:
>
> > @@ -585,39 +585,41 @@ function normestinv{T}(A::SparseMatrixCSC{T}, t::Integer = 2)
> >          S_old = copy(S)
> >          for j = 1:t
> >              for i = 1:n
> > -                S[i,j] = Y[i,j]==0?1:sign(Y[i,j])
> > +                S[i,j] = Y[i,j]==0?1:(Y[i,j]/abs(Y[i,j]))
>
> sign(z::Complex) is already defined this way, but is more efficient for
> non-complex types.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaLang/julia/pull/12592/files#r36921191>.
>

stevengj replied:
See also #12599.

marcusps replied:
Note that `sign(0) == 0` in `Base`, but this algorithm would require `sign(0)==1` (-1 would probably work fine too) [see the paper I referenced].

stevengj commented on base/sparse/linalg.jl:
```diff
@@ -585,39 +585,41 @@ function normestinv{T}(A::SparseMatrixCSC{T}, t::Integer = 2)
-                        S[i,j] = rand()>=0.5?1:-1
+                    if repeated
+                        for i = 1:n
+                            S[i,j] = rand()>=0.5?1:-1
```
`rand(-1:2:1)`?

StefanKarpinski replied:
Ditto, let's leave pre-existing code as is to keep this change as simple as possible. A separate PR to improve the code would be good too.

stevengj commented on base/sparse/linalg.jl:
```diff
@@ -585,39 +585,41 @@ function normestinv{T}(A::SparseMatrixCSC{T}, t::Integer = 2)
+        if T<:Real
+            # Check whether cols of S are parallel to cols of S or S_old
+            for j = 1:t
+                done = false
```
It would be cleaner to just use `while true` and a `break` statement.

StefanKarpinski replied:
Since this code was pre-existing, let's leave it as is in this PR and make tweaks to it separately. It would also be better to write `!done` and so on instead of the Matlabism `~done`.

Commit:
Fix complex matrix support in normestinv

+ Add tests
+ Remove check for parallel columns in complex matrices
+ Remove redundant capping of number of blocks
+ Add more robust default value for number of blocks
+ Minor change in complex "sign" for type stability

marcusps commented:
I renamed the pull request to be clearer, and fixed the problems that were causing tests to fail.

mfasi commented on base/sparse/linalg.jl:
```diff
@@ -532,10 +531,13 @@ function normestinv{T}(A::SparseMatrixCSC{T}, t::Integer = 2)
     end
```
This check seems redundant to me, I'd better delete it.

marcusps replied:
@mfasi Which check?

mfasi replied:
Sorry, I got the wrong line. I meant:
```javascript
     if t > n
         throw(ArgumentError("number of blocks must not be greater than $n"))
     end
```
It seems to me that we are guaranteed that `t` can't be greater than `n` at this point.

marcusps replied:
Ah, I was not sure whether to remove this or remove the `t = min(t,n)` line, but then decided that changing `t` silently was a bad idea.  I chose to remove that line and keep the check instead -- I also thought of changing the error to be a warning, and adding the `t = min(t,n)` back after the check.

Aside from thinking that changing silently `t` is a bad idea, I am happy with all other options.

mfasi replied:
I had mistaken the deletion for an insertion. The issue was pre-existing then (i.e. I introduced it), thank you for fixing it.

mfasi commented on base/sparse/linalg.jl:
```diff
@@ -585,42 +587,44 @@ function normestinv{T}(A::SparseMatrixCSC{T}, t::Integer = 2)
                 end
             end
-        end
+	end
```
Not sure the indentation is correct here, it does not seem aligned with what follows. Could you check it, please?

marcusps replied:
My bad -- new machine, editor was using tabs instead of spaces. Should be fixed now.

mfasi commented on base/sparse/linalg.jl:
```diff
@@ -585,42 +587,44 @@ function normestinv{T}(A::SparseMatrixCSC{T}, t::Integer = 2)
         for j = 1:t
             for i = 1:n
-                S[i,j] = Y[i,j]==0?1:sign(Y[i,j])
+                S[i,j] = Y[i,j]==0?one(Y[i,j]):sign(Y[i,j])
```
I'd rather put  `one(sign(Y[i,j]))`, to be consistent with the definition of  `sign(::Complex)`.

marcusps replied:
But `Y[i,j]` is never integer.  I define it to be to be `Float64` when the input is a subtype of `Real`, and Complex128 otherwise.

mfasi replied:
My bad this time. You're right, it's fine as it is.

marcusps replied:
No worries.

mfasi commented on base/sparse/linalg.jl:
```diff
@@ -532,10 +531,13 @@ function normestinv{T}(A::SparseMatrixCSC{T}, t::Integer = 2)
     ind_hist = Array(Int64, maxiter * t)
-    S = zeros(T, n, t)
+
+    Ti = T <: Real ? Float64 : Complex128
```
Maybe I overlooked something, but it seems to me that because of `est = zero(real(eltype(Y)))` we have two  issues:
* a complex value for a norm
* a double precision norm (estimate) for a single precision matrix
I think that the issue was already there and depended on the type declaration `est = zero(real(eltype(Y)))`, but I'm not sure what the best way to fix it may be.

marcusps replied:
`est = zero(real(eltype(Y)))` will still give a real value, so we will not be seeing a complex value for the norm.

As for the double precision, maybe
```julia
Ti = T <: Real ? typeof(float(zero(T))) : typeof(float(complex(zero(T))))
```
would do the trick?

mfasi replied:
It looks OK to me. Not sure we need to check whether `T` is a real type any more, though. I guess that `Ti = typeof(float(zero(T)))` would give the same `Ti` for complex and real types, which are the only ones we are interested in. A part from that, the PR seems ready to be merged.

marcusps replied:
True, that's much cleaner. I'll do that.

andreasnoack commented on base/sparse/linalg.jl:
```diff
@@ -532,11 +531,14 @@ function normestinv{T}(A::SparseMatrixCSC{T}, t::Integer = 2)
-    X = Array(T, n, t)
-    X[1:n,1] = 1
+    X = Array(Ti, n, t)
+    X[1:n,1] = one(Ti)
```
This is not really necessary because `1` will costlessly be converted to the element type of `X`.

andreasnoack commented on base/sparse/linalg.jl:
```diff
@@ -532,11 +531,14 @@ function normestinv{T}(A::SparseMatrixCSC{T}, t::Integer = 2)
     ind_hist = Array(Int64, maxiter * t)
-    S = zeros(T, n, t)
+
+    Ti = typeof(float(zero(T)))
```
What is the reasoning for this change? `S` is actually always integer-valued, right, so it could right as well be defined as a an integer matrix.

marcusps replied:
`S` is not always integer valued.  It is integer valued when `T` is Real, and it is complex otherwise -- that was  the problem I was running into after I added tests.

I can add a different type variable for `S` to catch that corner case.

andreasnoack replied:
I understand now. We probably don't have any cases where it would differ from `float`, but I guess `typeof(sign(zero(T)))` is the precise type, right?

andreasnoack commented:
@marcusps Is this one ready for merge now?

marcusps commented:
@andreasnoack I just pushed the minor changes suggested above.  The tests should pass (they pass locally), and at that point it should be a ready for a merge.

andreasnoack commented:
Thanks for fixing this.
